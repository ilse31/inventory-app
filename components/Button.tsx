import { colors, theme } from "@/constants/Colors";
import React from "react";
import {
  TouchableOpacity,
  Text,
  StyleSheet,
  ActivityIndicator,
  ViewStyle,
  TextStyle,
  StyleProp,
} from "react-native";

type ButtonVariant = "primary" | "secondary" | "outline" | "ghost" | "danger";
type ButtonSize = "small" | "medium" | "large";
type TextStyleArray = Array<TextStyle>;

interface ButtonProps {
  title: string;
  onPress: () => void;
  variant?: ButtonVariant;
  size?: ButtonSize;
  disabled?: boolean;
  loading?: boolean;
  icon?: React.ReactNode;
  style?: StyleProp<ViewStyle>;
  textStyle?: StyleProp<TextStyle>;
  fullWidth?: boolean;
}

export const Button: React.FC<ButtonProps> = ({
  title,
  onPress,
  variant = "primary",
  size = "medium",
  disabled = false,
  loading = false,
  icon,
  style,
  textStyle,
  fullWidth = false,
}) => {
  const getButtonStyle = () => {
    const baseStyle: ViewStyle = {
      ...styles.button,
      ...styles[`${size}Button`],
    };

    if (fullWidth) {
      baseStyle.width = "100%";
    }

    if (disabled) {
      return [
        baseStyle,
        styles[`${variant}Button`],
        styles.disabledButton,
        style,
      ];
    }

    return [baseStyle, styles[`${variant}Button`], style];
  };

  const getTextStyle = () => {
    const baseTextStyle: TextStyleArray = [
      { fontWeight: "600", textAlign: "center" },
      styles.text,
      styles[`${size}Text`],
      styles[`${variant}Text`],
    ];

    if (disabled) {
      baseTextStyle.push(styles.disabledText as TextStyle);
    }

    if (textStyle) {
      baseTextStyle.push(textStyle as TextStyle);
    }

    return baseTextStyle;
  };

  return (
    <TouchableOpacity
      style={getButtonStyle()}
      onPress={onPress}
      disabled={disabled || loading}
      activeOpacity={0.7}
    >
      {loading ? (
        <ActivityIndicator
          size="small"
          color={
            variant === "outline" || variant === "ghost"
              ? colors.primary
              : colors.text.light
          }
        />
      ) : (
        <>
          {icon && <>{icon}</>}
          <Text style={getTextStyle()}>{title}</Text>
        </>
      )}
    </TouchableOpacity>
  );
};

const styles = StyleSheet.create({
  button: {
    borderRadius: theme.borderRadius.md,
    flexDirection: "row",
    justifyContent: "center",
    alignItems: "center",
    gap: theme.spacing.sm,
  },
  text: {
    fontWeight: "600",
    textAlign: "center",
  },
  // Size variations
  smallButton: {
    paddingVertical: theme.spacing.xs,
    paddingHorizontal: theme.spacing.md,
  },
  mediumButton: {
    paddingVertical: theme.spacing.sm,
    paddingHorizontal: theme.spacing.lg,
  },
  largeButton: {
    paddingVertical: theme.spacing.md,
    paddingHorizontal: theme.spacing.xl,
  },
  smallText: {
    fontSize: 12,
  },
  mediumText: {
    fontSize: 14,
  },
  largeText: {
    fontSize: 16,
  },
  // Variant styles
  primaryButton: {
    backgroundColor: colors.primary,
  },
  secondaryButton: {
    backgroundColor: colors.secondary,
  },
  outlineButton: {
    backgroundColor: "transparent",
    borderWidth: 1,
    borderColor: colors.primary,
  },
  ghostButton: {
    backgroundColor: "transparent",
  },
  dangerButton: {
    backgroundColor: colors.danger,
  },
  primaryText: {
    color: colors.text.light,
  },
  secondaryText: {
    color: colors.text.light,
  },
  outlineText: {
    color: colors.primary,
  },
  ghostText: {
    color: colors.primary,
  },
  dangerText: {
    color: colors.text.light,
  },
  // Disabled state
  disabledButton: {
    opacity: 0.5,
  },
  disabledText: {
    opacity: 0.7,
  },
});
