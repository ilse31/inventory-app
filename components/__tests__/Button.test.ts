import React from "react";
import renderer from "react-test-renderer";
import { Text } from "react-native";
import { Button } from "../Button";

describe("Button component (react-test-renderer)", () => {
  it("renders correctly with title", () => {
    const tree = renderer
      .create(
        React.createElement(Button, { title: "Click Me", onPress: () => {} })
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders loading indicator when loading is true", () => {
    const tree = renderer
      .create(
        React.createElement(Button, {
          title: "Save",
          onPress: () => {},
          loading: true,
        })
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders as disabled when disabled prop is true", () => {
    const tree = renderer
      .create(
        React.createElement(Button, {
          title: "Disabled",
          onPress: () => {},
          disabled: true,
        })
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("renders with icon", () => {
    const Icon = () => React.createElement(Text, null, "icon");
    const tree = renderer
      .create(
        React.createElement(Button, {
          title: "With Icon",
          onPress: () => {},
          icon: React.createElement(Icon),
        })
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("applies fullWidth style", () => {
    const tree = renderer
      .create(
        React.createElement(Button, {
          title: "Full Width",
          onPress: () => {},
          fullWidth: true,
        })
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
