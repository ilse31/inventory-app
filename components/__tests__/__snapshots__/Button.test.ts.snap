// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component (react-test-renderer) applies fullWidth style 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#4A6FA5",
      "borderRadius": 8,
      "flexDirection": "row",
      "gap": 8,
      "justifyContent": "center",
      "opacity": 1,
      "paddingHorizontal": 24,
      "paddingVertical": 8,
      "width": "100%",
    }
  }
>
  <Text
    style={
      [
        {
          "fontWeight": "600",
          "textAlign": "center",
        },
        {
          "fontWeight": "600",
          "textAlign": "center",
        },
        {
          "fontSize": 14,
        },
        {
          "color": "#FFFFFF",
        },
      ]
    }
  >
    Full Width
  </Text>
</View>
`;

exports[`Button component (react-test-renderer) renders as disabled when disabled prop is true 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": true,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#4A6FA5",
      "borderRadius": 8,
      "flexDirection": "row",
      "gap": 8,
      "justifyContent": "center",
      "opacity": 0.5,
      "paddingHorizontal": 24,
      "paddingVertical": 8,
    }
  }
>
  <Text
    style={
      [
        {
          "fontWeight": "600",
          "textAlign": "center",
        },
        {
          "fontWeight": "600",
          "textAlign": "center",
        },
        {
          "fontSize": 14,
        },
        {
          "color": "#FFFFFF",
        },
        {
          "opacity": 0.7,
        },
      ]
    }
  >
    Disabled
  </Text>
</View>
`;

exports[`Button component (react-test-renderer) renders correctly with title 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#4A6FA5",
      "borderRadius": 8,
      "flexDirection": "row",
      "gap": 8,
      "justifyContent": "center",
      "opacity": 1,
      "paddingHorizontal": 24,
      "paddingVertical": 8,
    }
  }
>
  <Text
    style={
      [
        {
          "fontWeight": "600",
          "textAlign": "center",
        },
        {
          "fontWeight": "600",
          "textAlign": "center",
        },
        {
          "fontSize": 14,
        },
        {
          "color": "#FFFFFF",
        },
      ]
    }
  >
    Click Me
  </Text>
</View>
`;

exports[`Button component (react-test-renderer) renders loading indicator when loading is true 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": true,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={false}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#4A6FA5",
      "borderRadius": 8,
      "flexDirection": "row",
      "gap": 8,
      "justifyContent": "center",
      "opacity": 1,
      "paddingHorizontal": 24,
      "paddingVertical": 8,
    }
  }
>
  <ActivityIndicator
    color="#FFFFFF"
    size="small"
  />
</View>
`;

exports[`Button component (react-test-renderer) renders with icon 1`] = `
<View
  accessibilityState={
    {
      "busy": undefined,
      "checked": undefined,
      "disabled": false,
      "expanded": undefined,
      "selected": undefined,
    }
  }
  accessibilityValue={
    {
      "max": undefined,
      "min": undefined,
      "now": undefined,
      "text": undefined,
    }
  }
  accessible={true}
  collapsable={false}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    {
      "alignItems": "center",
      "backgroundColor": "#4A6FA5",
      "borderRadius": 8,
      "flexDirection": "row",
      "gap": 8,
      "justifyContent": "center",
      "opacity": 1,
      "paddingHorizontal": 24,
      "paddingVertical": 8,
    }
  }
>
  <Text>
    icon
  </Text>
  <Text
    style={
      [
        {
          "fontWeight": "600",
          "textAlign": "center",
        },
        {
          "fontWeight": "600",
          "textAlign": "center",
        },
        {
          "fontSize": 14,
        },
        {
          "color": "#FFFFFF",
        },
      ]
    }
  >
    With Icon
  </Text>
</View>
`;
